### PR TITLE
[vernac] duplicate extra dep is a warning

### PR DESCRIPTION
**Pull request by gares**
Before this patch the system would error if an Extra Dependency was found twice.
Unfortunately this is the case if the extra file is installed system wide, and you are compiling the same project.
AFAIK this affects only Coq-Elpi.

CC 8.16 RM @ppedrot since this may go (but not motivate) a minor release.

### CODE REVIEW

File: vernac/comExtraDeps.ml [hunk 1]
```diff
@@ -11,7 +11,14 @@
...
 
-let rec ensure_only_one_path_contains from file acc = function
+let warn_file_found_multiple_times =
+  CWarnings.create ~name:"ambiguous-extra-dep" ~category:"filesystem"
+    (fun (file,from,other,extra) ->
+      Pp.(str "File " ++ str file ++ str " found twice in " ++
+      DirPath.print from ++ str":" ++ spc () ++ str other ++ str " (selected)," ++
+      spc() ++ str extra ++ str ".") )
+
+let rec first_path_containing ?loc from file acc = function
```
Comment (ejgallego) [inline vernac/comExtraDeps.ml:21]: That went under my radar, but certain that this function doesn't belong here but on common code.

How is this code kept in sync with the one in coqdep?

Comment (gares) [reply vernac/comExtraDeps.ml:21]: Hum I guess you want me to use some code in tools/coqdep/lib.
The code there picks the first entry as well:
https://github.com/coq/coq/blob/2514b59568391079d0f3c362f6112b67fc380998/tools/coqdep/lib/common.ml#L228-L233

I did a quick look at the mli files, and I'm sure there is an API for this. Is it?

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 2]
```diff
@@ -775,6 +775,12 @@ This command has been available through the :cmd:`Comments` command,
 e.g. :n:`Comments From … Dependency …`.  The :n:`Comments` form is deprecated
 in Coq 8.16.
 
+.. warn:: File ... found twice in ...
+
+   The file is found in more than once in the top directories
+   associated to the given :n:`@dirpath`. In this case the first occurrence
+   is selected.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:782]:
```suggestion
   .. warn:: File ... found twice in ...

      The file is found in more than once in the top directories
      associated with the given :n:`@dirpath`. In this case the first occurrence
      is selected.
```

Please indent by 3 more spaces so it's under the `From .. Extra Dependency` command.

**Comment by ppedrot** [comment]
I think this is reasonable to backport this to 8.16.1.

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 2 above)
Comment (jfehrle) [reply doc/sphinx/proof-engine/vernacular-commands.rst]: Would also be nice to change the wording of the message to "more than once" to make it accurate if there are > 2 uses.  Up to you.

**Comment by gares** [comment]
I tried to use `coqdeplib` in vernac, and call `search_other_known` as `coqdep` does but I get:
```
File "vernac/dune", line 8, characters 28-37:
8 |  (libraries tactics parsing coqdeplib findlib.dynload))
                                ^^^^^^^^^
Error: Library "coqdeplib" is private, it cannot be a dependency of a public
library. You need to give "coqdeplib" a public name.
```
So I guess that library was not meant to be used outside coqdep.

@ejgallego what are you suggesting me to exactly?

**Comment by ejgallego** [comment]
> @ejgallego what are you suggesting me to exactly?

Indeed the library in coqdeplib is not for public use.

I think you want to put this functionality in `Loadpath`, as the plan we've been discussing is to split `Loadpath` to `Boot`, then have all the tools use it.

File: vernac/comExtraDeps.ml (hunk 1 above)
Comment (gares) [reply vernac/comExtraDeps.ml]: done

**Comment by gares** [comment]
code moved

**Comment by gares** [comment]
OK, this seems ready to me, but needs an assignee @coq/vernac-maintainers

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
will merge soon

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**